### PR TITLE
Feature/charts

### PR DESCRIPTION
**Pull request by arslanyaqoob7**
Charts for admin panel

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -39,6 +39,8 @@ gem 'cancancan', '3.0.0'
 gem 'wicked_pdf', '1.4.0'
 # For wicked pdf
 gem 'wkhtmltopdf-binary', '0.12.4'
+# GEM for states
+gem 'state_machines-activerecord'
```
Comment (aliadnanaslam7) [inline Gemfile]: specify version

File: app/controllers/admins_controller.rb [hunk 2]
```diff
@@ -4,6 +4,9 @@ class AdminsController < ApplicationController
 
   # GET /admin_dashboard
   def admin_dashboard
+    @sale_orders_list = @current_hospital.sale_orders.where(status: 'completed').group('year(created_at)').group('month(created_at)').sum(:total_sale_price)
+    @purchase_orders_list = @current_hospital.purchase_orders.Completed.group('year(created_at)').group('month(created_at)').sum(:total_cost_price)
+    @appointments_list = @current_hospital.appointments.group('year(start_time)').group('month(start_time)').group('day(start_time)').count
```
Comment (aliadnanaslam7) [inline app/controllers/admins_controller.rb]: move to model

Comment (aliadnanaslam7) [reply app/controllers/admins_controller.rb]: scope vs class methods

File: app/controllers/admins_controller.rb [hunk 3]
```diff
@@ -4,6 +4,9 @@ class AdminsController < ApplicationController
 
   # GET /admin_dashboard
   def admin_dashboard
+    @sale_orders_list = @current_hospital.sale_orders.where(status: 'completed').group('year(created_at)').group('month(created_at)').sum(:total_sale_price)
```
Comment (aliadnanaslam7) [inline app/controllers/admins_controller.rb]: Use constants

Comment (aliadnanaslam7) [reply app/controllers/admins_controller.rb]: use proper variable names

File: app/controllers/admins_controller.rb [hunk 4]
```diff
@@ -24,26 +27,26 @@ def edit
...
   def update
-    if @user.update(user_params)
-      respond_to do |format|
-        format.html { redirect_to admin_path(@user), success: t('user.updated_successfully') }
-      end
-    else
-      respond_to do |format|
-        format.html { redirect_to edit_admin_path, danger: @user.errors.full_messages.join(',') }
+    updated = @user.update(user_params)
+    updated ? flash[:success] = I18n.t('user.updated_successfully') : flash[:danger] = @user.errors.full_messages.join(', ')
```
Comment (aliadnanaslam7) [inline app/controllers/admins_controller.rb]: don't use I18n.t()

File: app/controllers/admins_controller.rb [hunk 5]
```diff
@@ -24,26 +27,26 @@ def edit
...
 
-  # Use callbacks to share common setup or constraints between actions.
   def set_user
     @user = User.find(params[:id])
   end
 
-  # Never trust parameters from the scary internet, only allow the white list through.
   def user_params
-    params.require(:admin).permit(:name, :email, :gender, :date_of_birth, :type)
+    params.require(:admin).permit(:name, :gender, :date_of_birth, :type)
```
Comment (aliadnanaslam7) [inline app/controllers/admins_controller.rb]: remove type and let admin edit email

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -66,7 +66,7 @@ def scope_sign_in
 
   def give_email_not_allowed_with_subdomain
     if signed_in?
-      redirect_to admin_dashboard_path if request.url.eql?(give_email_url(subdomain: request.subdomain))
+      redirect_to admin_dashboard_admins_path if request.url.eql?(give_email_url(subdomain: request.subdomain))
```
Comment (aliadnanaslam7) [inline app/controllers/application_controller.rb]: name should be dashboard_admins_path

File: app/controllers/departments_controller.rb [hunk 7]
```diff
@@ -51,28 +37,39 @@ def new
   def create
     @department.created_at = DateTime.now
```
Comment (aliadnanaslam7) [inline app/controllers/departments_controller.rb]: DateTime.now vs DateTime.current

File: app/controllers/departments_controller.rb [hunk 8]
```diff
@@ -51,28 +37,39 @@ def new
...
     @department.user_id = current_user.id
+    saved = @department.save
+    saved ? flash[:success] = t('departments.created_successfully') : flash[:danger] = @department.errors.full_messages.join(', ')
+    respond_to do |format|
+      if saved
+        format.html { redirect_to departments_path }
+      else
+        format.html { render new_department_path }
```
Comment (aliadnanaslam7) [inline app/controllers/departments_controller.rb]: check!

File: app/controllers/departments_controller.rb [hunk 9]
```diff
@@ -51,28 +37,39 @@ def new
...
-        end
+        format.html { render :edit }
       end
     end
   end
 
   # DELETE /departments/:id
   def destroy
+    destroyed = @department.destroy
+    destroyed ? flash[:success] = t('departments.destroyed_successfully') : flash[:danger] = @department.errors.full_messages.join(', ')
```
Comment (aliadnanaslam7) [inline app/controllers/departments_controller.rb]: use ```.destroyed?```

File: app/controllers/doctors_controller.rb [hunk 10]
```diff
@@ -37,50 +37,52 @@ def edit
     end
   end
 
-# PATCH/PUT /doctors/:id
+  # PATCH/PUT /doctors/:id
```
Comment (aliadnanaslam7) [inline app/controllers/doctors_controller.rb]: PATCH

File: app/controllers/doctors_controller.rb [hunk 11]
```diff
@@ -37,50 +37,52 @@ def edit
...
   def update
-    success_path = doctor_path(@doctor)
-    fail_path = edit_doctor_path
     previous_page = request.env['HTTP_REFERER']
-    success_path = admin_path(@doctor) if previous_page.include?('admin')
-    fail_path = previous_page if previous_page.include?('admin')
-
+    updated = @doctor.update(doctor_params)
+    updated ? flash[:success] = t('doctor.updated') : flash[:danger] = @doctor.errors.full_messages.join(', ')
```
Comment (aliadnanaslam7) [inline app/controllers/doctors_controller.rb]: Use best practices naming convention

File: app/controllers/doctors_controller.rb [hunk 12]
```diff
@@ -37,50 +37,52 @@ def edit
...
       end
     end
   end
 
-  # POST /doctors
-  def create
+  # DELETE /doctors/:id
+  def destroy
+    destroyed = @doctor.destroy
+    destroyed ? flash[:success] = t('doctor.destroyed') : flash[:danger] = @doctor.errors.full_messages.join(', ')
```
Comment (aliadnanaslam7) [inline app/controllers/doctors_controller.rb]: use ```destroyed?```

File: app/controllers/patients_controller.rb [hunk 13]
```diff
@@ -36,11 +37,15 @@ def show
 
   # DELETE /patients/:id
   def destroy
+    destroyed = @patient.destroy
     respond_to do |format|
-      if @patient.destroy
+      if destroyed
         format.html { redirect_to patients_path, success: t('patient.destroyed') }
```
Comment (aliadnanaslam7) [inline app/controllers/patients_controller.rb]: use flash

File: app/controllers/patients_controller.rb [hunk 14]
```diff
@@ -36,11 +37,15 @@ def show
...
       else
-        format.html { redirect_to patients_path, danger: @patient.errors.full_messages.join(',') }
+        format.html do
+          flash[:danger] = @patient.errors.full_messages.join(', ')
```
Comment (aliadnanaslam7) [inline app/controllers/patients_controller.rb]: flash.now?

File: app/controllers/patients_controller.rb [hunk 15]
```diff
@@ -61,17 +66,22 @@ def new
...
-    success_path = admin_path(@patient) if previous_page.include?('admin')
-    fail_path = previous_page if previous_page.include?('admin')
-
+    updated = @patient.update(patient_params)
+    updated ? flash[:success] = t('patient.updated') : flash[:danger] = @patient.errors.full_messages.join(', ')
     respond_to do |format|
-      if @patient.update(patient_params)
-        format.html { redirect_to success_path, success: t('patient.updated') }
+      if updated
+        if previous_page.include?('admin')
```
Comment (aliadnanaslam7) [inline app/controllers/patients_controller.rb]: :back

File: app/controllers/patients_controller.rb [hunk 16]
```diff
@@ -61,17 +66,22 @@ def new
...
+          format.html { redirect_to admin_path(@patient) }
+        else
+          format.html { redirect_to patient_path(@patient) }
```
Comment (aliadnanaslam7) [inline app/controllers/patients_controller.rb]: same view page

File: app/controllers/sale_orders_controller.rb [hunk 17]
```diff
@@ -62,7 +62,7 @@ def update
 
   # PATCH /sale_orders/:id/checkout
   def checkout
-    @sale_order.update(status: SaleOrder::STATUS[:complete])
+    @sale_order.checkout
```
Comment (aliadnanaslam7) [inline app/controllers/sale_orders_controller.rb]: state machine

File: app/models/ability.rb [hunk 18]
```diff
@@ -35,7 +36,7 @@ def initialize(user)
       can [:read, :complete], Appointment, doctor_id: user.id, hospital_id: Hospital.current_id
       can :create, Prescription, hospital_id: user.hospital_id, created_by: user.id
       can :add_medicine, Prescription, hospital_id: user.hospital_id
-      can :index, Prescription, Prescription.doctor_only(user) do |prescription|
+      can :index, Prescription, Prescription.unscoped.doctor_only(user) do |prescription|
```
Comment (aliadnanaslam7) [inline app/models/ability.rb]: Recheck

File: app/models/appointment.rb [hunk 19]
```diff
@@ -71,6 +71,6 @@ def valid_schedule?
     doctor_schedule = doctor.doctor_availabilities.find_by(day: start_time.strftime('%A'))
     return false if doctor_schedule.start_time.nil? || doctor_schedule.end_time.nil? || doctor_schedule.duration.nil?
 
-    start_time < end_time && (end_time - start_time) < doctor_schedule.duration.to_i && start_time.to_s.split[1] > doctor_schedule.start_time.to_s.split[1] && end_time.to_s.split[1] < doctor_schedule.end_time.to_s.split[1] && doctor.appointments.pending.where.not(id: id).where('(start_time BETWEEN ? AND ?) OR (end_time BETWEEN ? AND ?)', start_time, end_time, start_time, end_time).blank?
+    start_time < end_time && (end_time - start_time) < doctor_schedule.duration.to_time.to_i-'00:00'.to_time.to_i && start_time.to_s.split[1] > doctor_schedule.start_time.to_s.split[1] && end_time.to_s.split[1] < doctor_schedule.end_time.to_s.split[1] && doctor.appointments.pending.where.not(id: id).where('(start_time BETWEEN ? AND ?) OR (end_time BETWEEN ? AND ?)', start_time, end_time, start_time, end_time).blank?
```
Comment (aliadnanaslam7) [inline app/models/appointment.rb]: time zone

File: app/models/department.rb [hunk 20]
```diff
@@ -8,9 +8,9 @@ class Department < ActiveRecord::Base
   before_destroy :check_presence_of_doctors
 
   def check_presence_of_doctors
-    unless self.doctors.empty?
-      errors.add(:doctor_id, I18n.t('departments.can_not_destroyed'))
+    unless doctors.empty?
```
Comment (aliadnanaslam7) [inline app/models/department.rb]: early return?

File: app/models/doctor.rb [hunk 21]
```diff
@@ -1,18 +1,27 @@
...
+  validates :registration_number, uniqueness: { scope: :hospital_id }
+  validates :fee, presence: true
+  validates :fee, numericality: { greater_than: 0}
+  validate :type_doctor?
   accepts_nested_attributes_for :doctor_availabilities
 
   private
 
+  def type_doctor?
+    return true if type.eql?('Doctor')
```
Comment (aliadnanaslam7) [inline app/models/doctor.rb]: empty line

File: app/models/doctor_availability.rb [hunk 22]
```diff
@@ -20,12 +20,18 @@ def validate_start_and_end_time
   end
 
   def duration_in_range
-    if start_time.nil? && end_time.nil? && !duration.nil?
+    if !start_time.nil? && !end_time.nil? && duration.empty?
```
Comment (aliadnanaslam7) [inline app/models/doctor_availability.rb]: ```if/else```

File: app/models/doctor_availability.rb [hunk 23]
```diff
@@ -20,12 +20,18 @@ def validate_start_and_end_time
...
+      errors.add(:duration, I18n.t('doctor_availability.missing_duration'))
+    end
+    if start_time.nil? && end_time.nil? && !duration.eql?("")
       errors.add(:start_time, I18n.t('doctor_availability.missing_start_time'))
     end
-    unless duration.nil?
```
Comment (aliadnanaslam7) [inline app/models/doctor_availability.rb]: .present?

File: app/models/doctor_availability.rb [hunk 24]
```diff
@@ -20,12 +20,18 @@ def validate_start_and_end_time
...
-      unless (start_time.nil? && end_time.nil?)
-        errors.add(:duration, I18n.t('doctor_availability.duration')) if ((end_time - start_time) / 60) < ((duration.hour * 60) + duration.min)
+    if !duration.eql?('')
+      if !start_time.nil? && !end_time.nil?
+        duration_hours = duration.to_time.hour
```
Comment (aliadnanaslam7) [inline app/models/doctor_availability.rb]: .minute

File: app/models/sale_order.rb [hunk 25]
```diff
@@ -12,6 +12,11 @@ class SaleOrder < ActiveRecord::Base
   scope :patient_only, ->(user_id) { joins(prescription: :appointment).where(appointments: { patient_id: user_id, hospital_id: Hospital.current_id }) }
 
   STATUS = { draft: 'drafted', complete: 'completed' }.freeze
+  state_machine :status, :initial => :drafted do
+    event :checkout do
```
Comment (aliadnanaslam7) [inline app/models/sale_order.rb]: recheck

File: app/models/sale_order.rb [hunk 26]
```diff
@@ -12,6 +12,11 @@ class SaleOrder < ActiveRecord::Base
...
+      transition :drafted => :completed
```
Comment (aliadnanaslam7) [inline app/models/sale_order.rb]: new hash syntax

File: app/views/admins/admin_dashboard.html.erb [hunk 27]
```diff
@@ -126,4 +127,119 @@
...
+  <div class='row'>
+    <div class='col-md-6 bg-white rounded '>
+      <canvas id="sale_purchase_relation"></canvas>
+    </div>
+    <div class='col-md-6 bg-white rounded'>
+      <canvas id="appointments_chart"></canvas>
+    </div>
+  </div>
+  <script>
+    function change_days_name(days_numbers)
```
Comment (aliadnanaslam7) [inline app/views/admins/admin_dashboard.html.erb]: move to js file

File: app/views/admins/show.html.erb [hunk 28]
```diff
@@ -6,9 +6,11 @@
         <h6 class="text-primary font-weight-bold text-left col-4 py-2">Profile</h6>
         <div class="col-7"></div>
         <div class="col-1">
+          <% if can? :edit, @user %>
           <%= link_to edit_admin_path(@user), method: :get, class: "btn btn-outline-primary", 'data-toggle': 'tooltip', title: 'Edit' do %>
```
Comment (aliadnanaslam7) [inline app/views/admins/show.html.erb]: indentation

File: app/views/doctors/_form.html.erb [hunk 29]
```diff
@@ -1,84 +1,64 @@
...
+        <%= f.label :email %>
+        <%= f.text_field :email, class: 'form-control', required: :true %>
+      </div>
+      <div class="form-group">
+        <%= f.label :fee %>
+        <%= f.text_field :fee, class: 'form-control', required: :true %>
+      </div>
+      <div class="form-group">
+        <%= f.label :gender %>
+        <%= f.select :gender, [ 'Male', 'Female', 'Others'], { prompt: true }, class: 'form-control' %>
```
Comment (aliadnanaslam7) [inline app/views/doctors/_form.html.erb]: helper

File: app/views/doctors/_form.html.erb [hunk 30]
```diff
@@ -1,84 +1,64 @@
...
-          <%= f.label :registration_number %><br>
-          <%= f.text_field :registration_number, class: 'form-control', required: :true %>
+        <div class="form-group col-3">
+          <%= availability.time_field :end_time, class: 'form-control' %>
         </div>
-        <div class="form-group col-6">
-          <%= f.label :date_of_birth %><br>
-          <%= f.date_field :date_of_birth, class: 'form-control' %>
+        <div class="form-group col-3">
+          <%= availability.select :duration, ['00:15', '00:30', '00:45', '01:00'], { prompt: true }, class: 'form-control' %>
```
Comment (aliadnanaslam7) [inline app/views/doctors/_form.html.erb]: helper

File: app/views/doctors/_form.html.erb [hunk 31]
```diff
@@ -1,84 +1,64 @@
...
-      <div class="form-group">
-        <%= f.submit 'Submit', class: 'btn btn-primary form-control col-6' %>
-      </div>
-    <% end %>
-  </div>
-</center>
+      <% end %>
+    </div>
+    <div class="form-group">
+      <%= f.submit 'Submit', class: 'btn btn-primary form-control' %>
```
Comment (aliadnanaslam7) [inline app/views/doctors/_form.html.erb]: requests on double click

File: app/views/patients/index.html.erb [hunk 32]
```diff
@@ -3,12 +3,12 @@
   <div class="card shadow mb-4">
     <div class="card-header">
       <div class='row align-items-center'>
-        <div class='col-md-10'>
+        <div class='col-md-9'>
```
Comment (aliadnanaslam7) [inline app/views/patients/index.html.erb]: check on multiple devices

File: app/views/shared/_form.html.erb [hunk 33]
```diff
@@ -11,7 +11,7 @@
       </div>
       <div class="form-group">
         <%= f.label :gender %><br>
-        <%= f.select :gender, [ 'Male', 'Female', 'Others' ], {}, class: 'form-control' %></br>
+        <%= f.select :gender, [ 'Male', 'Female', 'Others' ], {}, class: 'form-control' %>
```
Comment (aliadnanaslam7) [inline app/views/shared/_form.html.erb]: helper

File: spec/models/doctor_spec.rb [hunk 34]
```diff
@@ -53,7 +61,9 @@
       doctor = Doctor.new(
         name: 'Demo Doctor',
         password: '12345678',
-        password_confirmation: '12345678'
+        password_confirmation: '12345678',
+        registration_number: 'h1d1',
```
Comment (aliadnanaslam7) [inline spec/models/doctor_spec.rb]: use before or let!

File: app/controllers/admins_controller.rb (hunk 4 above)
Comment (arslanyaqoob7) [reply app/controllers/admins_controller.rb]: removed

File: app/controllers/admins_controller.rb (hunk 5 above)
Comment (arslanyaqoob7) [reply app/controllers/admins_controller.rb]: removed

File: app/controllers/application_controller.rb (hunk 6 above)
Comment (arslanyaqoob7) [reply app/controllers/application_controller.rb]: changed

File: app/controllers/departments_controller.rb (hunk 8 above)
Comment (arslanyaqoob7) [reply app/controllers/departments_controller.rb]: removed

File: app/controllers/departments_controller.rb (hunk 9 above)
Comment (arslanyaqoob7) [reply app/controllers/departments_controller.rb]: changed

File: app/controllers/doctors_controller.rb (hunk 10 above)
Comment (arslanyaqoob7) [reply app/controllers/doctors_controller.rb]: changed

File: app/controllers/doctors_controller.rb (hunk 12 above)
Comment (arslanyaqoob7) [reply app/controllers/doctors_controller.rb]: changed

File: app/controllers/patients_controller.rb (hunk 13 above)
Comment (arslanyaqoob7) [reply app/controllers/patients_controller.rb]: changed

File: app/models/department.rb (hunk 20 above)
Comment (arslanyaqoob7) [reply app/models/department.rb]: changed

File: app/models/doctor.rb (hunk 21 above)
Comment (arslanyaqoob7) [reply app/models/doctor.rb]: added

File: app/models/doctor_availability.rb (hunk 22 above)
Comment (arslanyaqoob7) [reply app/models/doctor_availability.rb]: changed

File: app/models/doctor_availability.rb (hunk 23 above)
Comment (arslanyaqoob7) [reply app/models/doctor_availability.rb]: changed

File: app/models/sale_order.rb (hunk 26 above)
Comment (arslanyaqoob7) [reply app/models/sale_order.rb]: changed

File: app/views/doctors/_form.html.erb (hunk 30 above)
Comment (arslanyaqoob7) [reply app/views/doctors/_form.html.erb]: added

File: app/views/shared/_form.html.erb (hunk 33 above)
Comment (arslanyaqoob7) [reply app/views/shared/_form.html.erb]: added

File: app/views/patients/index.html.erb (hunk 32 above)
Comment (arslanyaqoob7) [reply app/views/patients/index.html.erb]: done

File: app/views/doctors/_form.html.erb (hunk 31 above)
Comment (arslanyaqoob7) [reply app/views/doctors/_form.html.erb]: send multiple requests

File: app/views/doctors/_form.html.erb (hunk 29 above)
Comment (arslanyaqoob7) [reply app/views/doctors/_form.html.erb]: added

File: app/views/admins/show.html.erb (hunk 28 above)
Comment (arslanyaqoob7) [reply app/views/admins/show.html.erb]: done

File: app/controllers/sale_orders_controller.rb (hunk 17 above)
Comment (arslanyaqoob7) [reply app/controllers/sale_orders_controller.rb]: removed

File: app/controllers/patients_controller.rb (hunk 14 above)
Comment (arslanyaqoob7) [reply app/controllers/patients_controller.rb]: done

File: app/models/sale_order.rb (hunk 25 above)
Comment (arslanyaqoob7) [reply app/models/sale_order.rb]: removed

File: app/controllers/doctors_controller.rb (hunk 11 above)
Comment (arslanyaqoob7) [reply app/controllers/doctors_controller.rb]: changed

File: Gemfile (hunk 1 above)
Comment (arslanyaqoob7) [reply Gemfile]: gem removed

File: app/controllers/patients_controller.rb (hunk 15 above)
Comment (arslanyaqoob7) [reply app/controllers/patients_controller.rb]: changed

File: app/controllers/patients_controller.rb (hunk 16 above)
Comment (arslanyaqoob7) [reply app/controllers/patients_controller.rb]: changed

File: app/controllers/admins_controller.rb (hunk 2 above)
Comment (arslanyaqoob7) [reply app/controllers/admins_controller.rb]: moved

File: app/controllers/admins_controller.rb (hunk 3 above)
Comment (arslanyaqoob7) [reply app/controllers/admins_controller.rb]: done

File: app/controllers/departments_controller.rb (hunk 7 above)
Comment (arslanyaqoob7) [reply app/controllers/departments_controller.rb]: changed

File: app/models/ability.rb (hunk 18 above)
Comment (arslanyaqoob7) [reply app/models/ability.rb]: removed

File: spec/models/doctor_spec.rb (hunk 34 above)
Comment (arslanyaqoob7) [reply spec/models/doctor_spec.rb]: changed

File: app/models/doctor_availability.rb (hunk 24 above)
Comment (arslanyaqoob7) [reply app/models/doctor_availability.rb]: changed

File: app/models/ability.rb (hunk 18 above)
Comment (arslanyaqoob7) [reply app/models/ability.rb]: removed

File: app/models/appointment.rb (hunk 19 above)
Comment (arslanyaqoob7) [reply app/models/appointment.rb]: issue resolved

**Pull request merged**